### ISSUE TITLE
Define a publishing workflow, possibly with `alr publish` support.

### ISSUE THREAD

**Issue by mosteo** [alire-project/alire#227]
For reference, the opam docs:
- https://opam.ocaml.org/doc/Packaging.html
- https://github.com/ocaml/opam-publish

I've received [here]() a reply about using tagged releases which is not entirely reassuring. Still, opam is using these releases. Let's hope the reply comes from an extra- person.

**Comment by mosteo** [comment]
Now I got a clear [answer](https://github.community/t5/How-to-use-Git-and-GitHub/Ensuring-identical-tarballs-from-releases/m-p/35808/highlight/true#M9217). For whatever reason they're keeping a escape clause open, so we should at least make a recommendation to generate your own tarball, and whether possible automate it.

**Comment by Fabien-Chouteau** [comment]
Should we consider hosting the archives in the index repo or another repo?

I don't know what the size limit is github.

**Comment by mosteo** [comment]
> Should we consider generating the archives with alr publish and hosting them in the index repo or another repo?

I don't think we can allow users to do so; at most we could help with creating the release/uploading the tarball to the user's repo (if it's in github). And we should be careful to avoid that we are spammed with automated PRs.

> I don't know what the size limit is github.

According to https://help.github.com/en/github/administering-a-repository/about-releases there's no limit on the global size of releases; only 2GB per file.

Also for reference, here's the release API: https://developer.github.com/v3/repos/releases/

**Comment by Fabien-Chouteau** [comment]
> I don't think we can allow users to do so;

Why not?

> at most we could help with creating the release/uploading the tarball to the user's repo (if it's in github).

Maybe with this indeed: https://developer.github.com/v3/repos/releases/

> And we should be careful to avoid that we are spammed with automated PRs.

I am not sure this is a problem considering that creating a PR requires a GitHub account. I think GitHub already has an anti spam account protection.

> > I don't know what the size limit is github.
> 
> According to https://help.github.com/en/github/administering-a-repository/about-releases there's no limit on the global size of releases; only 2GB per file.

I mean if we host the archives in a GitHub repo, as checked-in files. How big the repo can be? How many archives can we host?

**Comment by mosteo** [comment]
OK, I had misunderstood you.  [Here are the limits](https://help.github.com/en/github/managing-large-files/what-is-my-disk-quota): 100MB per file, 100GB per repo, but we enter the radar at 1GB. So, I don't think GitHub would look happily at expressly using a repo for binaries, even if we don't run afoul of any written rule. LFS quotas in the free tier (the 1GB BW) seem too restricted: https://help.github.com/en/github/managing-large-files/about-storage-and-bandwidth-usage

I guess to play it safe we should go for what they advise, which is using releases and uploading binaries there. That requires that users publish from a github repo, or directly from a tarball stored elsewhere.

Understood about the spam.

**Comment by Fabien-Chouteau** [comment]
I didn't know about this GitHub CLI tool called `hub`.

There is a command to create a release on a repo: https://hub.github.com/hub-release.1.html

This can be useful.

**Comment by TamaMcGlinn** [comment]
So, `alr publish [index]` from within an alire project should put a tar.gz file at the index location, and write/append to the toml file, such as he/helloworld.toml, correct?

**Comment by TamaMcGlinn** [comment]
For nuget, there is a nuget.spec file that defines which files go into the nuget package file. That saves the system from having hardcoded ignore lists; I understand we don't want the alire directory, but we can't reasonably know which other files (i.e. .gitignore, /bin/) should be excluded.

**Comment by mosteo** [comment]
> So, `alr publish [index]` from within an alire project should put a tar.gz file at the index location,

Actually the tarball should be put somewhere online, tarballs don't go into the index repository. Our current idea for automation is as a resource of a release in GitHub (unless the packager wants to override that) to ensure immutability.

>  and write/append to the toml file, such as he/helloworld.toml, correct?

Ideally, yes.

**Comment by mosteo** [comment]
> For nuget, there is a nuget.spec file that defines which files go into the nuget package file. That saves the system from having hardcoded ignore lists; I understand we don't want the alire directory, but we can't reasonably know which other files (i.e. .gitignore, /bin/) should be excluded.

Thanks for the pointer, I will check the nuget specs more closely. I understand `.nuget` files are maintained by hand? Do the Microsoft tools also help in populating these files?

**Comment by TamaMcGlinn** [comment]
I write .nuget files by hand, but there may be tools that help with this.

Where should the tarball be rooted? I.e. when I do `tar -xf [tarball]` should it immediately contain the .gpr file? Does alire extract to the packagename_version_filesystem directory, and then add that to GPR_PROJECT_PATH?

I looked at an example; the Ada PDF Writer package refers to a .zip file (so evidently alire supports zip as well?) that contains a single directory called apdf. Does that mean with statements will need to refer to 'apdf/some_gprfile.gpr' instead of some_gprfile.gpr directly?

**Comment by mosteo** [comment]
Your tarball/zipball must contain a single folder (like the `apdf` you mention), name not important, that in turn contains your sources.

> Does that mean with statements will need to refer to 'apdf/some_gprfile.gpr' instead of some_gprfile.gpr directly?

No, you must simply with `some_gpr.gpr` directly. When you define locations of gpr files in the manifest, you must give the complete relative path without including that archival top folder (that is, relative to your sources root previous to zipping).

You'll find more detailed instructions at [publishing.md](https://github.com/alire-project/alire/blob/master/doc/publishing.md) and [catalog-format-spec.md](https://github.com/alire-project/alire/blob/master/doc/catalog-format-spec.md)

**Comment by mosteo** [comment]
Completed as of 2.0
